app/services/attachments/concatenation_service.rb

```diff
@@ -0,0 +1,173 @@
+    # are deleted
+    def validate_and_concatenate(attachments, is_paired_end)
+      return unless attachments.length.positive?
+
+      validate_file_extensions(attachments)
+
+      if is_paired_end
+        validate_paired_end_files(attachments)
+        concatenate_paired_end_reads(attachments)
+      else
+        validate_single_end_files(attachments)
+        concatenate_single_end_reads(attachments)
+      end
+
+      # if option is selected then destroy the original files
+      attachments.each(&:destroy!) if concatenation_params[:delete_originals]
+    end
+
+    # Validates that the single end files are all the same type
+    def validate_single_end_files(attachments)
+      attachments.each do |attachment|
+        next unless attachment.metadata.key?('type')
+
+        raise AttachmentConcatenationError,
+              I18n.t('services.attachments.concatenation.incorrect_file_types')
+      end
+      true
+    end
+
+    # Validates that the paired end files are all the same type
+    def validate_paired_end_files(attachments)
+      attachments.each do |attachment|
+        next if attachment.metadata.key?('type') && attachments.first.metadata['type'] == 'illumina_pe'
+
+        raise AttachmentConcatenationError,
+              I18n.t('services.attachments.concatenation.incorrect_file_types')
+      end
+      true
+    end
+
+    # Validates if the file extensions all match for the attachments
+    def validate_file_extensions(attachments)
+      expected_extension = attachments.first.metadata['compression'] == 'gzip' ? '.gz' : '.fastq'
+      attachments.each do |attachment|
+        extension = attachment.file.filename.to_s.partition('.').last
+        next unless extension != expected_extension
+
+        raise AttachmentConcatenationError,
+              I18n.t('services.attachments.concatenation.incorrect_fastq_file_types')
+      end
+      true
+    end
+
+    # Concatenates the single end reads into a single-end file
+    def concatenate_single_end_reads(attachments)
+      basename = concatenation_params[:basename] || 'concatenated_file'
+      extension = attachments.first.metadata['compression'] == 'gzip' ? '.gz' : ''
+
+      blobs = retrieve_attachment_blobs(attachments)
+      composed_blob = ActiveStorage::Blob.compose(blobs, filename: "#{basename}_R1.fastq#{extension}")
+      Attachments::CreateService.new(current_user, attachable, { files: [composed_blob.signed_id] }).execute
+    end
+
+    # Concatenates the paired-end reads into a multiple paired-end files
+    def concatenate_paired_end_reads(attachments) # rubocop:disable Metrics/AbcSize
+      basename = concatenation_params[:basename] || 'concatenated_file'
+
+      extension = attachments.first.metadata['compression'] == 'gzip' ? '.gz' : ''
+
+      forward_reads, reverse_reads = attachments_to_paired_end_directional_reads(attachments)
+
+      files = []
+
+      blobs = retrieve_attachment_blobs(forward_reads)
```

Comment:
**nitpick (blocking):** I was thinking that the refactoring would be a method like `concatenate_attachments` and accept an array of attachments, and a filename. that way you could do the following:

```ruby
files << concatenate_attachments(forward_reads, "#{basename}_R1.fastq#{extension}").signed_id
```

This method would return the new `ActiveStorage::Blob`

Reply:
Refactored in 5e4fc6115a9cecf5130d295baaf4633dd3e1dcc4